### PR TITLE
System latency

### PR DESCRIPTION
**Pull request by leckijakub**
Created blog post about system latencies

### CODE REVIEW

File: blog/content/post/2020-08-19-system_latency.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+# Latency - The most crucial aspect of real-time systems
+
+## Intro
+
+Many words have been spoken about RT systems and their supremacy over typical systems in specific fields. But should we trust their name? Real-Time?\
+As usual, the answer is no. RT system just trying to be what they're called, that means, react to triggers like interruptions and response immediately. More often it's just not later than the strictly defined time, called latency.
+
+## Strategy
+
+So how can we measure the 'quality' of our rt system? Let's say we want a task to be weaken up after 200 us. To this task, we can use a timer which will generate interruption when the time runs out and wake up the task.
```
Comment (macpijan) [inline blog/content/post/2020-08-19-system_latency.md:24]:
```suggestion
So how can we measure the 'quality' of our rt system? Let's say we want a task to be woken up after 200 us. To this task, we can use a timer which will generate interruption when the time runs out and wake up the task.
```

File: blog/content/post/2020-08-19-system_latency.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+Before we put the task to sleep lets remember current time
```
Comment (macpijan) [inline blog/content/post/2020-08-19-system_latency.md:25]: is this finished sentence?

File: blog/content/post/2020-08-19-system_latency.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+To simulate a stressful environment for the system we're testing tools like [`hackbench`](http://manpages.ubuntu.com/manpages/xenial/man8/hackbench.8.html) and [`stressapptest`](http://manpages.ubuntu.com/manpages/trusty/man1/stressapptest.1.html) can be used.
+
+* [`Hackbench`](http://manpages.ubuntu.com/manpages/xenial/man8/hackbench.8.html) - tool for stressing kernel scheduler by creating pairs of threads communicating with each other via sockets
+
+* [`Stressapptest`](http://manpages.ubuntu.com/manpages/trusty/man1/stressapptest.1.html) - program for generating a realistic load of memory, CPU, and I/O by creating a specified amount of threads writing to memory, to file, or communicate with given IP server
+
+## Testing platform
+
+Tests will be performed on an i.mx8 platform with two builds of yocto-linux. A Regular version and a specially configured as RT using rt patch.
```
Comment (macpijan) [inline blog/content/post/2020-08-19-system_latency.md:47]: You can link to RP patch sources: https://mirrors.edge.kernel.org/pub/linux/kernel/projects/rt/

File: blog/content/post/2020-08-19-system_latency.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+## Tests cases
+
+To get a complete system's characteristic at least a couple of tests with a different load must be performed. To get fully reliable results, a test's time must take as least a couple of hours.
```
Comment (macpijan) [inline blog/content/post/2020-08-19-system_latency.md:51]:
```suggestion
To get a complete system's characteristic at least a couple of tests with a different load must be performed. To get fully reliable results, a test's time must take at least a couple of hours.
```

File: blog/content/post/2020-08-19-system_latency.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+* **_CASE 4:_** `cyclictest` with `stressapptest` on 4 memory threads testing 256MB of memory
+
+* **_CASE 5:_** `cyclictest` with `stressapptest` on 8 memory threads testing 256MB of memory, 4 I/O threads and 4 network threads
+
+## Results
+
+* **_CASE 1:_**(No load)
+_Regular Build_
+        <img src='../../static/img/system_latency_plots/normal/cyclic_alone.png' width=80%>
```
Comment (macpijan) [inline blog/content/post/2020-08-19-system_latency.md:70]: I'm not sure if those links will render correctly in the blog. Please look at other posts (the path needs to be changed most likely).

**Review by macpijan — CHANGES_REQUESTED** [review]
@leckijakub Please wrap lines at 80 chars as we suggest to do for documentation we prepare. See for yourself that reviewing is not that easy when lines are much longer and wrap: https://github.com/3mdeb/news-and-ideas/pull/313/files

File: blog/content/post/2020-08-19-system_latency.md (hunk 3 above)
Comment (leckijakub) [reply blog/content/post/2020-08-19-system_latency.md]: Link added

File: blog/content/post/2020-08-19-system_latency.md (hunk 5 above)
Comment (leckijakub) [reply blog/content/post/2020-08-19-system_latency.md:70]: Should it be like `img/system_latency_plots/rt/cyclic_alone.png`?

File: blog/content/post/2020-08-19-system_latency.md (hunk 2 above)
Comment (leckijakub) [reply blog/content/post/2020-08-19-system_latency.md:25]: Maybe like this?
```
Before we put the task to sleep lets get and remember the current time in some kind of variable.

1. The start point of measuring kernel latencies is when the interrupt occurs.
2. The First thing kernel must do is observe it and while there is a lot to do and to observe even that part can take some time.
3. After that, relevant ISR is called to handle interruption(in our case to wake up the task)
4. Next goes the kernel scheduler whose job is to manage all of the processes working in the system. When our task has been woken up, it was placed on the CPU queue indeed. So now it's the scheduler thing to handle the task ASAP.
5. When it's time come, CPU starts to process the task and we can end measuring the latency.

After that, we can once again get the current time, subtract it from the time before we put the task to sleep, and get the latency by subtracting time difference with the time given to timer.
```

**Comment by leckijakub** [comment]
Lines wrapped at 80 chars.

File: blog/content/post/2020-08-19-system_latency.md (hunk 5 above)
Comment (macpijan) [reply blog/content/post/2020-08-19-system_latency.md:94]: I believe the template post explains that: https://raw.githubusercontent.com/3mdeb/news-and-ideas/master/blog/content/post/YYYY-MM-DD-template-post.md

File: blog/content/post/2020-08-19-system_latency.md (hunk 2 above)
Comment (macpijan) [reply blog/content/post/2020-08-19-system_latency.md:31]: ok

File: blog/content/post/2020-08-19-system_latency.md (hunk 5 above)
Comment (leckijakub) [reply blog/content/post/2020-08-19-system_latency.md]: Fixed

File: blog/content/post/2020-08-19-system_latency.md [hunk 6]
```diff
@@ -0,0 +1,222 @@
+---
+author: jakub.lecki
```
Comment (artur-rs) [inline blog/content/post/2020-08-19-system_latency.md:2]: @leckijakub missing some crucial metadata in the header, check the template in the `blog/content/post/` directory

**Review by macpijan — APPROVED** [review]

File: blog/content/post/2020-08-19-system_latency.md (hunk 6 above)
Comment (leckijakub) [reply blog/content/post/2020-08-19-system_latency.md]: Missing metadata added

**Review by artur-rs — APPROVED** [review]

**Pull request merged**